File: Data Structures & Algorithms/count-paths/submission-1.java
class Solution {
    int [][] dp;
    public int uniquePaths(int m, int n) {
        dp = new int [m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                dp[i][j]=-1;
            }
        }
       return  rec(0,0,m,n);
    }
    public int rec(int i,int j,int m,int n){
        if(i<0 || i>=m || j>=n || j<0){
            return 0;
        }
        if(i==m-1 && j==n-1){
            return 1;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        return dp[i][j]= rec(i+1,j,m,n)+rec(i,j+1,m,n);

    }
}

File: Data Structures & Algorithms/reconstruct-flight-path/submission-1.js
class Solution {
    /**
     * @param {string[][]} tickets
     * @return {string[]}
     */
    findItinerary(tickets) {
        const adjMap = new Map();

        // 1. Build adjacency list sorting destinations alphabetically
        for (const [src, dst] of tickets) {
            if (!adjMap.has(src)) {
                adjMap.set(src, new PriorityQueue( (a, b) => a.localeCompare(b) ));
            }
            adjMap.get(src).enqueue(dst);
        }

        const res = [];

        // 2. Continuous Hierholzer Graph DFS Post-order traversal
        function rec(src) {
            const pq = adjMap.get(src);
            
            // While this airport still has valid outbound flights available
            while (pq && pq.size() > 0) {
                const nextDst = pq.dequeue(); // Pull out the next alphabetical airport
                rec(nextDst); // Recurse down
            }
            
            // Add to result only when we hit a terminal dead-end
            res.push(src);
        }

        // Start path traversal at JFK
        rec("JFK");

        // 3. Reverse the post-order collection to get the chronological itinerary path
        return res.reverse();
    }
}
